application/apps/text_grep/src/wc_reviewed.rs

```diff
@@ -0,0 +1,154 @@
+//Code reviewed on Friday with Dmitry.
+use std::io::Read;
+use std::sync::{Arc, mpsc};
+use std::collections::HashMap;
+use std::path::{Path, PathBuf};
+use std::thread;
+use grep_regex::RegexMatcher;
+use grep_searcher::Searcher;
+use grep_searcher::sinks::UTF8;
+use std::sync::mpsc::Sender;
+use std::time::Instant;
+use tokio_util::sync::CancellationToken;
+
+
+#[derive(Debug)]
+pub struct SearchResult {
+  pub file_path: String,
+  pub pattern_counts: HashMap<String, usize>,
+  pub error_message: Option<String>,
+}
+
+pub struct TextGrep;
+
+impl TextGrep {
+  pub fn new() -> Self {
+    TextGrep
+  }
+
+  pub async fn count_occurrences(
+    &self,
+    patterns: Vec<&str>,
+    file_paths: Vec<&str>,
+    chunk_size: usize,
+    cancel_token: CancellationToken,
+  ) -> Result<Vec<SearchResult>, String> {
+    let mut results = Vec::new();
+    let cancel_token_clone = cancel_token.clone();
+
+    let (sender, receiver) = mpsc::channel();
+    let (error_sender, error_receiver) = mpsc::channel();
+
+    let patterns_arc: Vec<_> = patterns.iter().map(|&p| Arc::from(p)).collect();
+    let file_paths_arc: Vec<_> = file_paths.iter().map(|&fp| Arc::from(fp)).collect();
+
+    let thread_handles: Vec<_> = file_paths_arc
+      .iter()
+      .map(|file_path| {
+        let patterns = patterns_arc.clone();
+        let sender = sender.clone();
+        let error_sender_clone = error_sender.clone();
+        let cancel_token = cancel_token_clone.clone();
+        let file_path = Arc::clone(file_path);
+        thread::spawn(move || {
+          if let Err(err) = process_file(&file_path, &patterns, chunk_size, &cancel_token, &sender) {
+            if error_sender_clone.send(err.to_string()).is_err() {
+              eprintln!("Error sending error message through channel");
+            }
+          }
+        })
+      })
+      .collect();
+
+    for handle in thread_handles {
+      handle.join().unwrap();
```

Comment:
`unwrap()` is acceptable only in tests and very-very specific use-cases. Please, check for error here and return some error if it is. For ex: `handle.join().map_err(|e| e.into())?;`